ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/ClientTest.java

```diff
@@ -0,0 +1,308 @@
+
+    assertThat(queryResult.columnNames(), is(DEFAULT_COLUMN_NAMES));
+    assertThat(queryResult.columnTypes(), is(DEFAULT_COLUMN_TYPES));
+
+    String queryId = queryResult.queryID();
+    assertThat(queryId, is(notNullValue()));
+    assertThat(server.getQueryIDs(), hasSize(1));
+    assertThat(server.getQueryIDs().contains(new PushQueryId(queryId)), is(true));
+  }
+
+  private void shouldDeliver(final Publisher<Row> publisher, final int numRows) {
+    TestSubscriber<Row> subscriber = new TestSubscriber<Row>() {
+      @Override
+      public synchronized void onSubscribe(final Subscription sub) {
+        super.onSubscribe(sub);
+        sub.request(numRows);
+      }
+    };
+    publisher.subscribe(subscriber);
+    assertThatEventually(subscriber::getValues, hasSize(numRows));
+    for (int i = 0; i < numRows; i++) {
+      assertThat(subscriber.getValues().get(i).values(), equalTo(rowWithIndex(i).getList()));
+    }
+    assertThat(subscriber.isCompleted(), equalTo(false));
+    assertThat(subscriber.getError(), is(nullValue()));
+  }
+
+  private void assertErrorWhen(
+      final Supplier<CompletableFuture<?>> queryRequest,
+      final String... errorMessages
+  ) throws Exception {
+    // Given
+    CountDownLatch latch = new CountDownLatch(1);
+
+    // When
+    final CompletableFuture<?> cf = queryRequest.get()
+        .exceptionally(error -> {
+
+          // Then
+          assertThat(error, notNullValue());
+          for (final String msg : errorMessages) {
+            assertThat(error.getMessage(), containsString(msg));
+          }
+
+          latch.countDown();
+          return null;
+        });
+    awaitLatch(latch, cf);
+  }
+
+  private static void awaitLatch(CountDownLatch latch, CompletableFuture<?> cf) throws Exception {
+    // Log reason for any failures, else output of failed tests is uninformative
+    cf.exceptionally(failure -> {
+      System.out.println("Failure reason: " + failure.getMessage());
+      return null;
+    });
+
+    awaitLatch(latch);
+  }
+
+  private static void awaitLatch(CountDownLatch latch) throws Exception {
+    assertThat(latch.await(2000, TimeUnit.MILLISECONDS), is(true));
+  }
+
+  private static class TestSubscriber<T> implements Subscriber<T> {
+
+    private Subscription sub;
+    private boolean completed;
+    private Throwable error;
+    private final List<T> values = new ArrayList<>();
+
+    public TestSubscriber() {
+    }
+
+    @Override
+    public synchronized void onSubscribe(final Subscription sub) {
+      this.sub = sub;
+    }
+
+    @Override
+    public synchronized void onNext(final T value) {
+      values.add(value);
+    }
+
+    @Override
+    public synchronized void onError(final Throwable t) {
+      this.error = t;
+    }
+
+    @Override
+    public synchronized void onComplete() {
+      this.completed = true;
+    }
+
+    public boolean isCompleted() {
```

Comment:
This and the rest of the methods should be synchronized as the results will be retrieved from a different thread.

Reply:
Done.